lib/streamlit/caching/cache_utils.py

```diff
@@ -228,6 +329,106 @@ def _show_cached_st_function_warning(
             dg.exception(e)
 
 
+"""
+Note [DeltaGenerator method invocation]
+There are 3 different ways an st function can be invoked:
+1. Implicitly on the main DG instance (plain `st.foo` calls)
+2. Implicitly on an active context's block (`st.foo` within a `with st.block` context)
+3. Explicitly on a DG instance (`st.sidebar.foo`, `my_container.foo`)
+
+To simplify replaying messages from a cached function result, we convert all of these
+to explicit invocations. How they get rewritten depends on both implicit vs explicit,
+and if the target DG is already known within the replay.
+
+Implicit invocation on a known DG -> Explicit invocation on that DG
+Implicit invocation on an unknown DG -> Rewrite as explicit invocation on main
+    This is situation 2 above, and the DG is a block entirely outside our function call,
```

Comment:
So far I understand that this is situation 2 but I'm confused what is main and how that interacts here.

Reply:
main is the implicit global DG. when you call `st.foo`, you are actually calling a method on main